### ISSUE TITLE
Batched translation speed is very slow compared to Fairseq

### ISSUE THREAD

**Issue by wenyong-h** [marian-nmt/marian-dev#266]
I'm benchmarking marian and fairseq.
The model i'm using is Transformer big, beam size is 3.
The gpu i'm using is p100 with cuda 9.0.
Deocder argument is setted following https://marian-nmt.github.io/docs/#batched-translation.

The result i'm getting is in the following table.
As you can see, the speedup is much worse compared to Fairseq as batch size increase.
The result i'm getting is so different from https://twitter.com/marian_nmt/status/1008476022478532608.
Do you know what might be wrong?

| decoder        | batch size           | wps  |
| ------------- |:-------------:| -----:|
| marian 1.4    | 1 | 101 |
| marian 1.5     | 1 | 104 |
| fairseq     | 1 | 88 | 
| marian 1.4   | 16 | 268 |
| marian 1.5   | 16 | 345 |
| fairseq         | 16 | 881 |
| marian 1.4   | 32 | 267 |
| marian 1.5   | 32 | 376 |
| fairseq         | 32 | 1482 |

**Comment by emjotde** [comment]
Hi,
The workspace setting here is for a much smaller model, I suppose it is re-allocating memory at the beginning and that is slow as hell. Try

```
./marian-decoder -m model.npz -v vocab.src.yml vocab.trg.yml -b 3 --normalize 0.6 \
    --mini-batch 64 --maxi-batch-sort src --maxi-batch 100 -w 10000 --max-length-factor 1.5
```

**Comment by wenyong-h** [comment]
I have already set workspace bigger than 10000.
But i didn't set max-length-factor, after set it to 1.5, 
with marian 1.5, batch size 32,  wps doubled, becomes 700.
but fairseq under the same settings, wps is 1550.

I did some profiling using nprof, found that , with Marian, GEMM is taking much bigger percent of time than Fairseq.  
Then i found Fairseq is using cuda 8.0,while Marian is using cuda 9.0
So maybe gemm in cuda 9.0 have some performance problem.
so, i'm going to first try Marian with cuda 8.0, see how it turns out.

**Comment by hieuhoang** [comment]
your analysis is probably correct, I've also found scalability issues in Marian, as discussed here
  https://github.com/marian-nmt/marian/issues/96
  https://arxiv.org/pdf/1805.09863.pdf
Probably due to the naive mini-batching implementation which should hopefully be sorted out soon.

As an extra big-up for my paper, it also has an optimization for your beam=3 scenario

**Comment by kpu** [comment]
Tensor cores?

**Comment by emjotde** [comment]
Something is weird. `--max-length-factor 1.5` gives me an improvement of maybe 10%, not 100% like in your case. Is your data somehow consisting of very long sentences? Character level maybe?

Can you share test set and model for benchmarking? 

@Hieu Not really comparable due to very different models and sizes. But indeed, I need to sort out the beam search.

**Comment by emjotde** [comment]
@kpu since Fairseq is using CUDA 8.0, unlikely.

**Comment by wenyong-h** [comment]
@emjotde The model i'm benchmarking is only trained for about 5 hours, so it might decode forever.   I tried a fully trained model（converted from model trained in tensorflow）. Then i get same result as you, with  `--max-length-factor 1.5`, only 10% improvement.

**Comment by wenyong-h** [comment]
It's my coworker who did the conversion.  He did the conversion very quick, so I  guess he was just collecting the corresponding weight matrix from Tensorflow model graph, then save it to the numpy npz file using the names of marian model.

**Comment by wenyong-h** [comment]
As for my test setting, i'm using a single P100 for testing.
I have analysis the sentence lengths in my data, after sort,  the length difference in each batch is very small, usually by just one or two.
So the not-fully-optimized beam-search unlikely to be the main reason.

**Comment by emjotde** [comment]
Hm, is the converted model actually able to create translations? That would be a surprise. I would not expect to be fully compatible.
Anyway, both conditions seem to be slightly broken, and not quite suitable for benchmarking yet.

**Comment by wenyong-h** [comment]
Well, the company i'm working for have strict rules about sharing code to the public.
But i can ask my coworker to share his experience in detail.

**Comment by emjotde** [comment]
These are WPS numbers I am getting on a TitanX Pascal for the same en-de system I used for the Volta benchmarks. What puzzles me is, that you report only a 3x speed-up going from 1 to 16 and then nothing from going beyond that, while I have 4x for 16, 6x for 32 and then 8x for 128. It flattens out above 128.

| batch | 1.4 | master |
|-------|-----|----|
|    1 |  158 |   164|
|  16 |  534 |   681|
|  32 |  683 |   961|
|  64 |  723 | 1180 |
| 128|  748 | 1249 |

**Comment by wenyong-h** [comment]
My original WPS numbers have problems, as i'm using models trained only a few hours ,without setting the max-length-factor.
I redo the test with cuda 8, and set max-length-factor to 1.5.
This is what  i got

| decoder        | batch size           | wps  |max-length-factor
| ------------- |:-------------:| -----:|---------:|
|master|1|106|1.5|
|master|16|540|1.5|
|master|32|769|1.5|
|master with fully trained model|32|710|default|
|master with fully trained model|32|780|1.5|

As you can see, the speedup is very similar to yours.
I guess your are testing Transformer base, while i'm testing Transformer big, so my wps is smaller.

Anyway,  The speedup is much worse than Fairseq,  and cuda version is not the problem.

**Comment by wenyong-h** [comment]
I add some log in fairseq to see how much computation is saved by removing finished sentence during decoding.
I did the calculation , `sum_of_batch_size_of_each_decode_step/ (max_batch_size * number_of_decode_steps)`.
What I got is 0.686.
So by the removing-finished-sentence optimization, Fariseq could save about 31.4% computation.
Even if this would lead to 31.4% speedup(which is unlikely),  still cann't explain the problem,  as Fariseq is 100% quick than marian when batch size is 32.

**Comment by emjotde** [comment]
I am decoding with a 813 MB Transformer big model, not base. 
Yes, my guess was that improved beam search would result in about 25-30% improvement. 

What's the average sentence length in source tokens in your test set?

**Comment by wenyong-h** [comment]
I thought P100 was better than TitanX, am I wrong?
The average sentence length is about 15, I don't remember clearly, I will check that when i got back to work.

Or maybe if you got time, you could try Fairseq on your machine, with the same dataset,

**Comment by kpu** [comment]
Pascal Titan X is generally faster than P100.  Also, it may not matter but Pascal Titan X is compute level 6.1 whereas P100 is compute level 6.

**Comment by hieuhoang** [comment]
@wenyong-h Do you know how much time is spent on decoding v. encoding? I assume you only remove completed sentences during decoding, not encoding.

**Comment by wenyong-h** [comment]
@hieuhoang I didn't measure that. But I think decoding takes far more time than encoding, 
because,  
1 ,every decode step needs to go through the decoder self attention, encoder-decoder attention layers,  while encoder need to go through self attention layers only once
2 only decoder need beam search, 
3 decoder have a huge softmax to compute.

For you second question, yeah, Fairseq remove completed sentences during decoding, not encoding.

**Comment by emjotde** [comment]
@hieuhoang They are doing exactly what I described we should do in this issue #175. There isn't much else than can be done here. 

I checked again the numbers on the Volta, there I am getting the 110 to 115 sentences per second. Seems we handle different GPUs worse than Fairseq.

**Comment by emjotde** [comment]
@wenyong-h Since you mentioned the softmax, how big is your target vocabulary?

**Comment by wenyong-h** [comment]
target vocabulary is about 60k

**Comment by hieuhoang** [comment]
@emjotde ok, just making sure there's nothing else going on. 

can't tell from the issue what happens when beam>1. Will you still keep number of hypos the same for all uncompleted sentence, or follow amun and drop completed hypos?

**Comment by emjotde** [comment]
I was going to drop completed sentences, so keep them around until they are done. Currently the beamsize is a max over all sentences in a batch, so it is decreasing. What's missing is the case where we drop a sentence from the batch when beam reaches 0. We go down to at most beam=1. 

I do not think you can easily have different beam sizes for different items in a batch. if you want to keep it general.

**Comment by emjotde** [comment]
Dropping the finished sentences from the encoder is basically like pretending that your batch never contained your sentence in the first place, so you can then go back to the original algorithm without modifications and do again max over remaining beams.

**Comment by hieuhoang** [comment]
Wonder what fairseq does

**Comment by emjotde** [comment]
Did not look on the code, but I would guess just that. That should already be pretty good. 

We are losing a lot of compute in cases where a stray long sentences causes all other 120 sentences to be computed with beam 1. The overhead for garbage is then more than 90%.

**Comment by emjotde** [comment]
And I would guess it gets worse the larger the vocabulary. So the larger vocabulary here (60000 vs my 32000) might very well explain why we cannot reproduce the beneficial speed of the Volta results.

**Comment by wenyong-h** [comment]
I modified the code of marian and Fairseq, removed the logic of "remove finised sentence or beam from batch" from both, and force the deocder always decode the same number of step as the length of input sentence. and use test set consist of sentences with the same length 15.
The following is test result.

| batch size        | marian         | fairseq  |
| ------------- |:-------------:| -----:|
| 1 | 118 | 112 |
| 8 | 601 | 770 |
| 16 | 910 | 1235 |
| 32 | 1328 | 1686 |
| 64 | 1729 | 1911 |


As you can see, even without the remove-finised-sentence optimization, Fairseq still perform better when batch size get bigger.
So I guess maybe some gpu op of marian is not as performent as pytorch, and that must be batch matrix multipy used to implement multi-head attention or the output layer.
So I take the code  used to implement  multi-head attention from both framework to do some benchmark.
Below is the code, and the results.

code for Marian:
```c++
int beam_size = 3;
std::vector<int> batch_sizes = { 1, 8, 16, 32, 64 };
for (auto batch_size : batch_sizes) {
    std::cout << "batch size: " << batch_size << std:endl;
    auto g = New<ExpressionGraph>(false, false);
    g->setDevice({ 1, DeviceType::gpu });
    g->reserveWorkspaceMB(8512);
    for (int i = 0; i < 7; ++i) {
        g->clear();
        auto q = g->constant({ batch_size * beam_size, 16, 1, 64 }, inits::glorot_uniform);
        auto k = g->constant({ batch_size * beam_size, 16, 15, 64 }, inits::glorot_uniform);
        g->forward();
        boost::timer::auto_cpu_timer timer;
        float scale = 0.125;
        auto z = bdot(q, k, false, true, scale);
        g->forward();
    }
}
```
code for Fairseq:
```python
for batch_size in [1, 8, 16, 32, 64]:
  print(f'batch size: {batch_size}')
  q = torch.cuda.FloatTensor(batch_size * 16 * 3, 1, 64) 
  k = torch.cuda.FloatTensor(batch_size * 16 * 3, 64, 15) 
  for i in range(7): 
    torch.cuda.synchronize() 
    s = timer() 
    u = torch.bmm(q, k) 
    torch.cuda.synchronize() 
    print(timer() - s)
```
average time used to execute a bdot or bmm, unit of time used in the table is 1e-5 sec

| batch size        | marian         | fairseq  |
| ------------- |:-------------:| -----:|
| 1 | 4.2 | 2.4 |
| 8 | 5.6 | 2.6 |
| 16 | 6.8 | 3.3 |
| 32 | 9.8 | 4.3 |
| 64 | 14.8 | 6.3 |


So it seems my guess may be right, 
though I'm  not sure if I'm doing the benchmark right, correct me if I'm wrong.

**Comment by wenyong-h** [comment]
Hi, I'll list what i found so far.
1 cudaMalloc in CopyRows is big performance hit for Transformer. Avoid it could give me a 250 wps increase when batch size is 32, see #273 
2 beam search of marian have to copy things between gpu and cpu, while Fairseq is does not
3 In fairseq, multi-head projection of q,k,v is done in single matrix multiply. if q, k, v is projected from same source.
if not that case, then at least k,v is  projected from same source, so k, v projection will be done in a single matrix multiply.   While in marian, q,k,v projection is done with three matrix multiply sequentially.

**Comment by emjotde** [comment]
Time to revisit this.

**Comment by emjotde** [comment]
I believe this is solved now.
